### PR TITLE
Rewrite the chapter on subtyping and variance

### PR DESCRIPTION
**Pull request by conradludgate**
This is to address the issues raised in #339

### CODE REVIEW

File: src/subtyping.md
Comment (mejrs) [inline src/subtyping.md:12]:
```suggestion
Subtyping is the idea that one type can be used in place of another.
```

I'm not happy with the current line, because a reader probably doesn't know what a subtype is either.

File: src/subtyping.md
Comment (mejrs) [inline src/subtyping.md:29]: I think the [`Error`](https://doc.rust-lang.org/std/error/trait.Error.html) trait probably fits better here.

- It already exists in the standard library.
- It's a trait users will have used (or at least seen), and maybe even implemented.

When I see people asking about "traits requiring other traits" it is usually the example I reach for, because it's very obvious - In order to use something as an error, it must be displayable..

File: src/subtyping.md
Comment (Nilstrieb) [inline src/subtyping.md:41]:
```suggestion
    println!("a = {a:?} b = {b:?}");
```
This raises the MSRV of the nomicon examples, but I think we can take that :). Same for the other examples.

File: src/subtyping.md
Comment (mejrs) [inline src/subtyping.md:37]: Personally I would lead the chapter with an example, *before even mentioning variance and subtyping*, showing an example that should compile but doesn't, where the user can see that it makes sense, and then drop *we can fix this with variance*

**Review by Nilstrieb — COMMENTED** [review]
I really liked the rewrite, it's shorter and clearer, explaining everything without cats and dogs.

Small suggestion: I'd use the cool new formatting syntax

File: src/subtyping.md
Comment (mejrs) [inline src/subtyping.md:74]: I actually really liked this section, can we keep it? I think it's important to say this stuff is important when writing unsafe. And maybe include an example of messing it up?

**Review by mejrs — COMMENTED** [review]
Thanks, this is much better :)

File: src/subtyping.md
Comment (conradludgate) [reply src/subtyping.md:74]: Absolutely. It must have slipped past me while I was cleaning it up

File: src/subtyping.md
Comment (conradludgate) [reply src/subtyping.md:181]: Can you think of a good way to explain that?

The original mostly just says that since `UnsafeCell<T>` allows you to treat a `&T` as a `&mut T`, it should have the same invariance

File: src/subtyping.md
Comment (nrc) [inline src/subtyping.md:10]:
```suggestion
Let's start with an example.
```

File: src/subtyping.md
Comment (nrc) [inline src/subtyping.md:27]:
```suggestion
In a conservative implementation of lifetimes, since `a` and `b` have differing lifetimes,
```

File: src/subtyping.md
Comment (nrc) [inline src/subtyping.md:50]: I'm not sure if requirements is the best word here. Perhaps it is better to use terminology like "A value which has type Sub can be used in any way that a value of type Super can, and possibly more"

File: src/subtyping.md
Comment (nrc) [inline src/subtyping.md:139]: Variance is not just a Rust concept

File: src/subtyping.md
Comment (LegionMammal978) [inline src/subtyping.md:48]:
```suggestion
Let's define that `Sub` is a subtype of `Super` (we'll be using the notation `Sub: Super` throughout this chapter).
```

Comment (conradludgate) [reply src/subtyping.md:48]: Aren't these two the same thing? `'a` outlives `'b` means that `'a` is a subtype of `'b`

File: src/subtyping.md
Comment (LegionMammal978) [inline src/subtyping.md:53]:
```suggestion
An example of simple subtyping that exists in the language is [supertraits](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-supertraits-to-require-one-traits-functionality-within-another-trait):
```
It might also be useful to put the URL at the end using an anchor, as in `send-and-sync.md`.

File: src/subtyping.md
Comment (LegionMammal978) [inline src/subtyping.md:89]:
```suggestion
e.g. `&'static str` is a subtype of `&'b str`, then we can let them coerce,
```

File: src/subtyping.md
Comment (LegionMammal978) [inline src/subtyping.md:110]:
```suggestion
It's not always as simple as this example, though. To understand that, let's try to extend this example a bit:
```

File: src/subtyping.md
Comment (LegionMammal978) [inline src/subtyping.md:247]:
```suggestion
This is counter to the `&T` case:
```

File: src/subtyping.md
Comment (conradludgate) [reply src/subtyping.md:42]: Hmm. This is contradictory to how the original page represents `&'a T` as a type constructor with type arguments `'a` and `T`

File: src/subtyping.md
Comment (nrc) [reply src/subtyping.md:48]: I don't think outlives is a subtype relation because lifetimes are not types, it could be called sub-lifetimes, but we don't call it that.

Comment (jackh726) [reply src/subtyping.md:48]: > Aren't these two the same thing? `'a` outlives `'b` means that `'a` is a subtype of `'b`

That's actually the opposite of what's true: `'a: 'b` means `'a <: 'b`

**Review by jyn514 — APPROVED** [review]
I like the idea of using lifetimes to start instead of the cat + dog examples :) I only skimmed the PR, but it would be nice to have an example of why `fn borrow<'a>(x: &'a Chunk<'a>) {}` works (usually) but `fn borrow_mut<'a>(x: &'a mut Chunk<'a>) {}` is almost impossible to use. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1eed7dc7b916058c4a06a7da908717e4

File: src/subtyping.md
Comment (danielhenrymantilla) [inline src/subtyping.md:72]:
```suggestion
`'a` defines a region of code.
```

When part of a borrow, it thus represents a region of code within which a value is borrowed. A fortiori, the region within which that borrowee is alive needs to be at least as big as the region of the borrow.

In other words, _lifetime_ is a misnomer; while it could have made sense back in the day with lexical lifetimes (and even then it would have been a stretch), now with NLL it doesn't really match either the life-time of the borrowee, nor that of the borrower:

```rs
/* 0 */ {
/* 1 */     let mut x = 42;
/* 2 */     {
/* 3 */         let r = &x;
/* 4 */         println!("{}", *r);
/* 5 */         x = 0;
/* 6 */         println!("{}", x);
/* 7 */     }
/* 8 */     println!("{}", x);
/* 9 */ }
```

  - The _alive_-time of `x` is `1.. 9`
  - The _alive_-time of `r` is `3.. 7`
  - The _lifetime_ / span of the borrow of the `&x` borrow (held by `r`) is `3.. 5`

Before NLL, the second and third spans had to match; they don't since.

File: src/subtyping.md
Comment (danielhenrymantilla) [inline src/subtyping.md:74]:
```suggestion
`'long : 'short` if and only if `'long` defines a region of code that **completely contains** `'short` (_i.e._, `'long ⊇ 'short`)
```

I've added the whitespace to `'a : 'b` kind of bounds, since we're not really within rustfmt-ed code anymore, we're teaching stuff, and `:` for lifetimes represents an (antisymmetrical) binary "boolean" operator, so there is no reason to favor the left hand side over the right hand side.

File: src/subtyping.md
Comment (danielhenrymantilla) [inline src/subtyping.md:86]:
```suggestion
Going back to our example above, we can say that `'static : 'b`, as mentioned just above.
```

or, if my previous suggestions were not to be added:

```suggestion
Going back to our example above, we can say that `'static: 'b`. Indeed `'static` stands for `'forever`: the never-ending / infinite region. Since any other region is then thus contained within it (`'static ⊇ 'b`), we have `'static : 'b` for any lifetime `'b`.
```

File: src/subtyping.md
Comment (danielhenrymantilla) [inline src/subtyping.md:76]:
```suggestion
`'long` may define a region larger than `'short`, but that still fits our definition.

### Some basic properties

For any lifetimes `'a`, `'b`, and `'c`:

  - `'a : 'a` (a region fits inside itself / `'a = 'a ⇒ 'a ⊇ 'a`);

  - if `'a : 'b` and `'b : 'a`, then `'a = 'b` (the only way for two regions to be inside one another is if they're equal / `('a ⊇ 'b & 'b ⊇ 'a) ⇒ 'a = 'b`);

  - if `'a : 'b`, and `'b : 'c`, then `'a : 'c` (if `'b` is part of `'a`, and `'c` is part of `'b`, then necessarily `'c` is part of `'a` / `('a ⊇ 'b & 'b ⊇ 'c) ⇒ 'c ⊇ 'a`.

  - `'static` stands for `'forever`, the never-ending / infinite region. Thus `'static ⊇ 'a` for any `'a`; that is:
      `'static : 'a`.
```

This serves both as concrete pedagogical examples, and genuine tools to be used later on 🙂

File: src/subtyping.md
Comment (danielhenrymantilla) [inline src/subtyping.md:81]:
```suggestion
And unless you write unsafe code, the compiler will automatically handle all the corner cases for you.

  - Note that the aforementioned properties (`:` between lifetimes defines a partial order with `'static` being the maximum) remain true even the "region" representation falls apart.
```

File: src/subtyping.md
Comment (danielhenrymantilla) [inline src/subtyping.md:88]:
```suggestion
For now, let's also accept the idea that subtypes of lifetimes can be passed through references
_i.e._, "the lifetime of borrow is allowed to shrink" (more on this in [Variance](#variance)),
```

File: src/subtyping.md
Comment (danielhenrymantilla) [inline src/subtyping.md:89]: _coerce_ is not really a great word, here, since it does have a specific meaning in Rust which subtyping is distinct from.

 -  Maybe something along the lines of?

    ```suggestion
    _e.g._, `&'static str` is a subtype of `&'b str`, then we can let a `&'static str` "become" a `&'b str`.
    ```

    🤷

File: src/subtyping.md
Comment (danielhenrymantilla) [inline src/subtyping.md:109]:
```suggestion
Above, we glossed over the fact that `'static : 'b` implied that `&'static T <: &'b T`. This uses a property known as _variance_.
```

File: src/subtyping.md
Comment (danielhenrymantilla) [inline src/subtyping.md:136]:
```suggestion
This means that `&mut &'static str`**cannot** be a *subtype* of `&mut &'b str`,
```

File: src/subtyping.md
Comment (danielhenrymantilla) [inline src/subtyping.md:139]: Could we avoid type constructors and talk of _generic types_ instead? It ought to be clearer. Moreover, it incidentally allows to add the sentence: talking of the variance of a non-generic type makes no sense.

  - We're glossing over `fn(&str) <: fn(&'concrete str)` here, by the way, but I think that's fine

File: src/subtyping.md
Comment (danielhenrymantilla) [inline src/subtyping.md:156]: FWIW, I agree with @conradludgate's point about the advantage of using `:` everywhere; it allows to use this "subtyping is passed through" / "subtyping is inverted" wave-handed but very useful mnemonic.

I may also be useful to draw a tangential (footnote?) parallel with function monotonicity in mathematics, since for those that know that notion they can suddenly realize they know what all this is all about 🙂

**Review by danielhenrymantilla — APPROVED** [review]
I would very much not mix traits into this chapter ⚠️, but other than that, I'm quite fond of this PR 🙂 ✅

File: src/subtyping.md
Comment (jyn514) [reply src/subtyping.md:89]: "degrade"  (like atoms) might be a more descriptive work, or "reduce".

**Comment by conradludgate** [comment]
Thanks everyone for the reviews. It seems overall people are in favour of this general form but there's a bit of bike shedding over syntax.

I'll make all the grammatical and language changes suggested

File: src/subtyping.md
Comment (jackh726) [inline src/subtyping.md:48]: Just to make sure this doesn't get lost: I don't think we should use this notation. `Sub <: Super` is correct. `Sub: Super` is too confusing with outlives relationship, and generally the opposite of what is (potentially) expected.

Comment (conradludgate) [reply src/subtyping.md:48]: Yeah these syntax suggestions are growing on me. I think i agree with them now and I will perform the changes

Comment (mejrs) [reply src/subtyping.md:48]: Why not use existing symbols like `⊂` for this, rather than inventing our own?

Comment (jackh726) [reply src/subtyping.md:48]: `<:` is well-used to indicate subtyping and is easy to type.

Comment (danielhenrymantilla) [reply src/subtyping.md:48]: I think that `<:` seems to be too widespread/common in this scenario not to use it. That being said, when generic types are involved, things start to be quite noise, angle-bracket-wise:

```console
Result<Vec<&'static str>> <: Result<Vec<&'_ str>>
```

So `⊂:`, besides being a nicer smiling face 😄, could be a decent middle-ground with @mejrs's proposal:

```console
Result<Vec<&'static str>> ⊂: Result<Vec<&'_ str>>
```

But, it would have the sad side-effect that it wouldn't be that nice in conjunction with `⊇` for regions, and I _really_ find `⊇` for regions to be an incredibly useful notation:

```rust
for<'lt>
      // 'static ⊇ 'lt, and thus:
    &'static str ⊂ &'lt str
,
```

So, if we were not to use `<:`, I'd rather we use a notation with _motion_ to represent subtyping:

```rust
for<'lt>
      // 'static ⊇ 'lt, and thus:
    &'static str → &'lt str
,
Result<Vec<&'static str>> → Result<Vec<&'_ str>>
```

Something like this is, imho, the most readable notation for people unacquainted with type theory (which, let's not forget it, is the target audience for this kind of chapters).

**Comment by jackh726** [comment]
@conradludgate the types team (wg-traits on zulip) is planning to go over this PR next Friday at 9 am Eastern Time. You're welcome to join if you'd like!

**Review by nikomatsakis — COMMENTED** [review]
Gave this a read. Some notes:

* I didn't spot any errors, beyond a certain imprecision in distinguishing types/traits/lifetimes and their relationships.
* I've not read the rest of the rustonomicon so I can't really comment on how consistent that usage is, but I think we are definitely going to have to change it eventually. In particular, I think it's we want to have a "sublifetime" relationship (let's write it `<:_l` for now). Even though `'a <:_l 'b iff 'a: 'b`, that doesn't make outlives equivalent, because it's also valid to have an outlives relation between lifetimes and types (and even, in the work I've been doing lately, types and types). 
* That said, I'm not clear 100% on the goals of this book/chapter, so I don't know how much it makes sense to worry at this moment about the distinctions there.
* What did seem to be missing from this chapter was a bit of practical advice: how do you (for example) write a test that variance is what you expect? You don't mention how the compiler infers variance and the fact that it has associated semver/soundness implications, but that seems important.

File: src/subtyping.md
Comment (lcnr) [inline src/subtyping.md:180]: don't need the first row anymore as you're removing the `*`

File: src/subtyping.md
Comment (lcnr) [inline src/subtyping.md:75]:
```suggestion
_e.g._ `&'static str` is a subtype of `&'world str`, then we can "downgrade" `&'static str` into a `&'world str`.
```

File: src/subtyping.md
Comment (lcnr) [inline src/subtyping.md:79]:
```suggestion
fn debug<'a>(a: &'a str, b: &'a str) {
```
to stay consistent with the original example

File: src/subtyping.md
Comment (lcnr) [inline src/subtyping.md:229]: there is no `&` here?

Maybe actually stay with `&'a str` instead of using some type parameter `T` here

File: src/subtyping.md
Comment (lcnr) [inline src/subtyping.md:175]: that's just wrong, higher rank lifetimes don't work independently of subtyping.

you don't have to fix that here as it's an existing issue

**Review by lcnr — COMMENTED** [review]
a few nits, then r=me.

there are also some other open reviews you may want to deal with before merging

File: src/subtyping.md
Comment (rib) [reply src/subtyping.md:43]: > Nit: subtyping applies to types, not lifetimes

I'm not sure that's really a nitpick. I think it's potentially the crux of why this chapter explaining lifetime variance has been quite unintuitive (at least for me) - since it tries too hard to compare lifetime variance to subtyping which is a different thing.

I've tried to expand on this here: https://github.com/rust-lang/nomicon/issues/339#issuecomment-1422694052 as well as link back to other issues that discussed how rust should define variance for lifetimes.

**Comment by jyn514** [comment]
oof, it's been a while ... what's the current state of this PR?

I would prefer to merge this even if it's not 100% complete rather than have it hang in limbo; we can always improve it in follow-ups and my understand is it's already more accurate than the existing docs.

**Review by ehuss — APPROVED** [review]
Thanks @conradludgate for working on this!

I went through and applied some of the review suggestions. If anyone has anything more specific that they would like to see, feel free to open a new PR to update this.

**Pull request merged**
